### ISSUE TITLE
Ch. 4: Exercise 12

### ISSUE THREAD

**Issue by IvanKuchin** [ageron/handson-ml#452]
Hello Aurelien, 

 I did ex.12 on my own and had problem with troubleshooting. 
In my case early stop works 1 out of 10 times and stopped after less than 20 iterations (IMO it depends on random initialization). That behavior doesn't seems correct to me. After 2 days of troubleshooting I rebuilt the code so that it can be compared line-by-line against yours. And turns out that early stop to works following conditions must be met:

1. implementation MUST have regularization. (partial derivatives have no regularization part in the book)
2. pair (learning rate and regularization hyperparameter) must be big enough (0.1, 0.1). If it would be (0.01, 0.1) , (0.1, 0.01) and so on even 5M iterations didn't hit the early stop. 

May be worth to point out those caveats in exercise statement.

**Comment by ageron** [comment]
Good catch, thanks a lot @IvanKuchin , I'll add a note.
